src/autopas/containers/LeavingParticleCollector.h

```diff
@@ -9,38 +9,76 @@
 
 #include "autopas/options/IteratorBehavior.h"
 #include "autopas/particles/OwnershipState.h"
+#include "autopas/utils/ArrayMath.h"
 #include "autopas/utils/inBox.h"
 
-namespace autopas {
 /**
  * Namespace to collect leaving particles from a container.
  */
-namespace LeavingParticleCollector {
+namespace autopas::LeavingParticleCollector {
 /**
  * Collects leaving particles and marks halo particles as dummy.
  * @note This function does not move or actually delete any particles!
  * @tparam ContainerType The type of the container.
  * @param container The container from which particles should be collected.
  * @return Returns a vector of leaving particles.
- * @todo: openmp parallelization, general optimizations (do not iterate over all particles, but only over boundary,
- * eventually merge both loops?
  */
 template <class ContainerType>
 std::vector<typename ContainerType::ParticleType> collectParticlesAndMarkNonOwnedAsDummy(ContainerType &container) {
-  // First mark halo particles as dummy!
-  for (auto iter = container.begin(autopas::IteratorBehavior::halo); iter.isValid(); ++iter) {
-    iter->setOwnershipState(OwnershipState::dummy);
-  }
+  using namespace autopas::utils::ArrayMath::literals;
+
+  const auto interactionLength = container.getInteractionLength();
+  const auto containerDimensions = container.getBoxMax() - container.getBoxMin();
+  const auto lowerHaloCorner = container.getBoxMin() - interactionLength;
+  const auto upperHaloCorner = container.getBoxMax() + interactionLength;
+  // We need to extend these boundaries into the container because we need to find particles that are stored in the
+  // container but have since moved out. Extending by skin is not enough since we are interested in particles that have
+  // travelled beyond that and skin can be 0.
```

Comment:
I think this could be clearer. For example, why are these particles not caught by the region iterator? Is this extension by `interactionLength` determined by something or just a heuristic?

Reply:
This is how all of our containers work. Since we rebuild every container not in every iteration but when verlet-rebuild-frequency is reached all subcontainers (neighborlists, cells, ...) are extended by the skin.

Reply:
No, my point was not about the extension by skin. 

I'm going to explain my understanding of the function. Correct me if I am wrong.

All particles marked halo need to be deleted. It is not simply enough to iterate over regions outside of the box since a halo particle may have moved inside the box. So that we are not iterating over the whole container, we need some heuristic for the most a particle might have moved into the container. We can't use verlet skin /2 because we might have skin=0 (i.e. we rebuild every iteration) or, in the case of Luis' dynamic verlet lists, it is not necessary to use a verlet skin such that no particle should ever move more than verlet skin / 2. So you propose `interactionLength` as a heuristic. 

It makes total sense to have this heuristic as at least cutoff because if a particle could travel further than that in one iteration then we could have particles jumping over other particles when there should be a repulsive force between them at short distances (which implies that a smaller step-size should be used anyway). This heuristic makes a lot of sense in the types of particle simulations we are currently dealing with. But at the end of the day, this is still some heuristic - and is potentially not valid for every possible simulation. 

If I am correct in my understanding then, to prevent future headaches, this should be highlighted in the comments and the reasoning for the heuristic should be provided. You could go further and provide a "debug mode" whereby the region iterator is replaced with a normal iterator and a warning is printed if a halo particle is found outside of the regions of the region iterators. But this wouldn't be necessary for this PR and an issue could be opened. Finally, is there some reason for `interactionLength` instead of cutoff?

If I am not correct in my understanding then provide clearer reasoning to the extension by `iteractionLength`.

Reply:
In general your understanding is correct. Just some clarifications:
- What we add/subtract from the box boundaries is not a heuristic but an upper bound for how far particles have moved inside of the container. If this is not the lowest upper bound possible, it is still correct just not as efficient as it could be.
- Theoretically, the upper bound for movement per timestep is `skinPerTimestep/2`. If `skin = 0` this doesn't help us. However, having a `skin > cutoff` also makes no sense from a performance perspective, hence `cutoff` seems a reasonable upper bound.
- Yes cutoff should be more reasonable than the full interaction length. Don't ask why I had `interactionLengt` there first -> c78a8c683b390537c95c5a617222e231ea8b502a